### PR TITLE
Issue #69 - Adding local Foreign Key Constraints to CSV-W

### PR DESCRIPTION
**Pull request by robons**
I've added the foreign key constraints to the CSV-W we output. I had to re-work slightly how we output code-lists.

I also realised that I'd completely forgotten to add primary keys to the CSV-Ws so I added those to both the dataset and code-list CSV-Ws with some behaviour tests to boot.

This satisfies issue #69.

### CODE REVIEW

**Review by canwaf — APPROVED** [review]
- [x] Check behave tests (rdf.py has new steps) 'the RDF should not contain any instances of {entity_type}'
- [x] Validate code is correct.
- [x] Run unit and behave tests locally
- [x] Ask questions about techniques I don't understand
- [x] Confirm approaches which may need fixing outside the scope of this ticket

File: csvqb/csvqb/tests/behaviour/qbwriter.feature [hunk 1]
```diff
@@ -21,4 +21,39 @@ Feature: Test outputting CSV-Ws with Qb flavouring.
...
+    And csv2rdf on "some-qube.csv-metadata.json" should succeed
+    And the RDF should not contain any instances of "http://www.w3.org/2004/02/skos/core#ConceptScheme"
+
+  Scenario: A QbCube should fail to validate where foreign key constraints are not met.
+    Given a single-measure QbCube named "Some Qube" with codes not defined in the code-list
+    When the cube is serialised to CSV-W
+    Then the file at "a-code-list.csv-metadata.json" should exist
+    And csvlint validation of "some-qube.csv-metadata.json" should fail with "unmatched_foreign_key_reference"
+
+  Scenario: A QbCube with existing dimensions should not do foreign key checks.
```
[the remark this answers is not in the document]
Comment (robons) [reply csvqb/csvqb/tests/behaviour/qbwriter.feature:40]: Yep, but I want to make sure we don't currently break anything when a user uses existing dimensions.

File: csvqb/csvqb/tests/unit/writers/test_qbwriter.py [hunk 2]
```diff
@@ -474,67 +473,89 @@ def test_virtual_columns_generated_for_multi_meas_obs_val():
...
+            "Local Dimension",
+            NewQbDimension.from_data("Name of New Dimension", data["Local Dimension"]),
+        ),
+        QbColumn(
+            "Value",
+            QbSingleMeasureObservationValue(
+                ExistingQbMeasure("http://example.com/measures/existing_measure"),
+                NewQbUnit("New Unit"),
+            ),
+        ),
```
[the remark this answers is not in the document]
Comment (robons) [reply csvqb/csvqb/tests/unit/writers/test_qbwriter.py:508]: Black.

[the remark this answers is not in the document]
Comment (canwaf) [reply csvqb/csvqb/tests/unit/writers/test_qbwriter.py:508]: 👍🏻

File: csvqb/csvqb/tests/behaviour/qbwriter.feature (hunk 1 above)
[the remark this answers is not in the document]
Comment (canwaf) [reply csvqb/csvqb/tests/behaviour/qbwriter.feature:40]: Is there a ticket for this yet?

[the remark this answers is not in the document]
Comment (robons) [reply csvqb/csvqb/tests/behaviour/qbwriter.feature:40]: The closest we have is #73 - but we'd need somewhere to be able to pull the code-list's CSV-Ws from. PMD doesn't currently allow us to do that, and we don't have anywhere we can really look this information up from either.

**Pull request merged**